### ISSUE TITLE
Confused by usage of `cross_entropy` in XGB Crab tutorial

### ISSUE THREAD

**Issue by xiaodaigh** [alan-turing-institute/MLJTutorials#26]
`cross_entropy` is a loss function. Basically, the higher it is the worse the model.

In
https://alan-turing-institute.github.io/MLJTutorials/pub/end-to-end/crabs-xgb.html

I see that it says the model is doing a good job at 100 rounds

![image](https://user-images.githubusercontent.com/4497189/70025236-83746880-15f0-11ea-93e9-f736b2aa04b9.png)

But shouldn't it say that 10 is the best because it has the lowest Cross Entropy? Is the entropy in the chart the average of average entropy of each record in the test set in the k-fold?

**Comment by tlienart** [comment]
Thanks for raising this,

@ablaom could you comment on this please?, it's true that this should not look that way; initially  I thought that it's `learning_curve!` that was re-scaling the measurements so that the curves would look the same irrelevant of whether it was a score or a loss but I don't see this in the implementation of `learning_curve!` so there's something else going on

**Comment by ablaom** [comment]
Thanks @xiaodaigh for bringing this to our attention. I agree, the learning curve looks suspicious. However, at the end of my investigation, I have concluded it is actually okay. Please read on. 

For context, here are learning curves for two other loss functions, which look more "normal":

![hinge](https://user-images.githubusercontent.com/30517088/70404439-f4a79600-1a9e-11ea-858e-6d20f3eaa07f.png)

![mr](https://user-images.githubusercontent.com/30517088/70404447-025d1b80-1a9f-11ea-87e3-e1371d24d548.png)

I also looked at the BrierScore(), which, as a score, is higher the "better" the prediction:

![breir](https://user-images.githubusercontent.com/30517088/70404503-333d5080-1a9f-11ea-98a3-3cc11a0eaf31.png)

So why is cross_entropy qualitatively so different?

First note that it is theoretically possible for cross_entropy to go UP when misclassification rate (say) goes down. I can give you a simple counterexample. The point is that cross entropy has the following characteristic (which is not always desirable, despite its popularity): The loss is *very* sensitive to low probabilistic predictions for what is actually observed (compared to high probabilities), much more so than the other measures mentioned. 

We should also keep in mind that *the metric begin minimised by XGBoostClassifier is* **not** *cross_entropy*, but rms, as I understand it (or perhaps it is rms applied to a "raw" target value before the logistic map is applied to get the probabilistic prediction?).

To check if the cross-entropy increase could be explained by the appearance of low probabilities in the test, I plotted two histograms of the probabilities (for what was observed) for num_round=1  and num_round=500 and discovered the appearance of some significantly lower probabilities in the latter case:

![histograms](https://user-images.githubusercontent.com/30517088/70404910-587e8e80-1aa0-11ea-988d-5721242f97a0.png)

See here ---^

Given that I have I also rechecked the cross_entropy function tests, I am happy with this explanation. That is, I do not have reason to believe there is a problem with `cross_entropy`.

@xiaodaigh Are you happy with this explanation?

@tlienart In any event, to avoid confusion, we should perhaps replace `cross_entropy` in the tutorial with HingeLoss() or BrierScore().

**Comment by tlienart** [comment]
Thanks for taking the time to  write this Anthony, I'll update the metric to BriersScore and might add some more stuff stolen from your comment here.

**Comment by xiaodaigh** [comment]
> @xiaodaigh Are you happy with this explanation?

I think that's a good technical overview, but the central issue I have is:

* Is the plot plotting average cross_entropy? 
  * If it is then the tutorial should say that 10 rounds is better than 100 rounds. In fact, for such a small problem, we should be searching between 1 and 10 rounds
  * if it is not, then be clear what it's plotting or how `learning_curve!` transform cross_entropy before plotting

I think from the explanation that it is plotting average `cross_entropy` on the test set. So I suggest changing the search to 1 to 10 rounds.

**Comment by ablaom** [comment]
> * If it is then the tutorial should say that 10 rounds is better than 100 rounds. In fact, for such a small 

Yes, that is clear. And we should show more detail for n < 10 . But, for the purposes of a tutorial, we should probably pick a measure that XGBoostRegressor actually improves, at least in some initial period. 

* Is the plot plotting average cross_entropy? 

Not sure I understand this question. What do you mean by "average" here? Do you mean aggregated over the test set, or some other average? Ahh. Perhaps you mean we are locally averaging over `num_round`? No, we are not.

**Comment by xiaodaigh** [comment]
> What do you mean by "average" here?

I thought because it's doing a k-fold (I believe the default is 6-fold) validation, it must be computing the `cross_entropy` k-times, so for each `nround` value there are 6 computations of `cross_entropy`, which needs to be averaged to be compared?

**Comment by ablaom** [comment]
Yes, indeed, with `resampling=CV()` the metric is aggregated across (six) folds. This raises a separate issue, which is that usually, for an iterative model and big datasets, you will probably want to use the default resampling strategy, which is a `Holdout`, for generating this kind of learning curve. From the learning_curves doc string:

---
If using a `Holdout` `resampling` strategy, and the specified
hyperparameter is the number of iterations in some iterative model
(and that model has an appropriately overloaded `MLJBase.update`
method) then training is not restarted from scratch for each increment
of the parameter, ie the model is trained progressively.

---

That said,  XGBoost.jl does *not* implement `MLJBase.update` (which it should, for this very reason).

**Comment by tlienart** [comment]
So as a note, I came back to this today hoping to  fix this and realised something was just off (the cross entropy shouldn't  give  a wholly different  answer and the numerical susceptibility is not the culprit). So I tested against sklearn and it turns out our implementation of the cross entropy loss is just wrong see https://github.com/alan-turing-institute/MLJBase.jl/issues/147

Thanks a lot @xiaodaigh for spotting this.

Note also that, as initially  expected, more trees is better in most normal metrics it points to 300-500 trees.
